sklearn/linear_model/stochastic_gradient.py

```diff
@@ -233,6 +244,38 @@ def _allocate_parameter_mem(self, n_classes, n_features, coef_init=None,
                                                dtype=np.float64,
                                                order="C")
 
+    def _train_validation_split(self, X, y, sample_weight):
+        """Split the dataset between training set and validation set
+
+        Returns
+        -------
+        validation_set : array, shape(n_samples, )
+            Equal to 1 on the validation set, 0 on the training set
+        """
+        n_samples = X.shape[0]
+        if not self.early_stopping:
+            # use the full set for training, with an empty validation set
+            return np.zeros(n_samples, dtype=np.int16)
+
+        tmp = train_test_split(X, y, np.arange(n_samples), sample_weight,
+                               test_size=self.validation_fraction,
+                               random_state=self.random_state)
+        X_train, X_val, y_train, y_val = tmp[:4]
+        idx_train, idx_val, sample_weight_train, sample_weight_val = tmp[4:8]
+
+        self._X_val = X_val
```

Comment:
should we delattr these at the end of fitting?

Reply:
Done